content/algorithms/dag/index.md

```diff
@@ -0,0 +1,216 @@
+such that following those directions will never form a closed loop.
+
+A directed graph is a DAG if and only if it can be topologically ordered
+by arranging the vertices as a linear ordering that is consistent with all edge directions.
+
+## Topological sort
+
+### Example
+
+```{code-cell} ipython3
+# get a topological sort of a dag
+print(list(nx.topological_sort(clothing_graph)))
+```
+
+### Applications
+
+The canonical application of topological sorting is in scheduling a sequence of jobs
+or tasks based on their dependencies.
+The jobs are represented by vertices, and there is an edge from $u$ to $v$
+if job $u$ must be completed before job $v$ can be started
+(for example, when washing clothes, the washing machine must finish before we put the clothes in the dryer).
+Then, a topological sort gives an order in which to perform the jobs.
+
+A closely related application of topological sorting algorithms
+was first studied in the early 1960s in the context of the PERT technique for scheduling in project management.
+In this application, the vertices of a graph represent the milestones of a project,
+and the edges represent tasks that must be performed between one milestone and another.
+Topological sorting forms the basis of linear-time algorithms for finding
+the critical path of the project, a sequence of milestones and tasks that controls
+the length of the overall project schedule.
+
+In computer science, applications of this type arise in instruction scheduling,
+ordering of formula cell evaluation when recomputing formula values in spreadsheets,
+logic synthesis, determining the order of compilation tasks to perform in makefiles,
+data serialization, and resolving symbol dependencies in linkers.
+It is also used to decide in which order to load tables with foreign keys in databases.
+
+### Definition
+
+A topological sort of a directed acyclic graph $G = (V, E)$ is a linear ordering of all its vertices
+such that if $G$ contains an edge $(u, v)$, then $u$ appears before $v$ in the ordering.
+
+It is worth noting that if the graph contains a cycle, then no linear ordering is possible.
+
+It is useful to view a topological sort of a graph as an ordering of its vertices
+along a horizontal line so that all directed edges go from left to right.
+
+### Kahn's algorithm
+
+First, find a list of "start nodes" which have no incoming edges and insert them into a set S;
+at least one such node must exist in a non-empty acyclic graph. Then:
+
+```
+L <- Empty list that will contain the sorted elements
+S <- Set of all nodes with no incoming edge
+
+while S is not empty do
+    remove a node n from S
+    add n to L
+    for each node m with an edge e from n to m do
+        remove edge e from the graph
+        if m has no other incoming edges then
+            insert m into S
+
+if graph has edges then
+    return error  # graph has at least one cycle
+else 
+    return L  # a topologically sorted order
+```
```

Comment:
Pseudocode is okay, but a working (and readable) Python implementation would be preferable!